### ISSUE TITLE
Better utilize concurrency when pulling form submissions

### ISSUE THREAD

**Issue by batkinson** [opendatakit/briefcase#138]
This is [an issue that was previously raised on the Google Group with submission download performance](https://groups.google.com/d/msg/opendatakit/kzuRmSSPkfg/N6fgFDodyHEJ). However, I stumbled upon this independently and after doing some investigation on it myself, would ask that we reopen the discussion.

Specifically, I believe that adding support for concurrent submission downloads would improve  performance considerably in environment with _high latency_, regardless of the available bandwidth. Mitch's assessment did not talk about the design of the download protocol itself. Over high-latency network connections, the time waiting for previous requests to complete can easily dominate the time necessary to complete a pull.

My basic understanding of the protocol, for each form being pulled:
- Fetch the form definition, the media manifest, and files listed in the manifest
- Fetch a chunk of 100 (max) submission ids for the form, starting from the last cursor value
- One-by-one fetch each submission, including the media files

Consider an example pull:

- A single form definition with 10 media files
- A significant number of submissions, let's say 1000
- Each submission has an average of 1.5 attached media files
- The client's round-trip latency averages 350 milliseconds
- Each http request is serviced and sent instantaneously, eliminating bandwidth from the discussion
- Assume TCP connection time is near-constant with regards to the number of requests using HTTP 1.1 persistent connections

Because each request is performed serially, the time it will take to complete the pull is the round-trip time multiplied by the number of http requests:

- 1 form definition
- 1 form manifest
- 10 media files
- 10 id chunks containing 100 submission ids each (covering all 1000 submissions)
- 1000 form submissions
- 1500 submission media files

That works out to be 2,522 round trips. Multiplying that by our 350ms round trip time, and even with infinite processing speed on the server and infinite bandwidth, the pull takes: 2,522 requests * 350ms = 882,700 ms (approximately 14.7 minutes).

My reason for investigating this was a pull of a single form definition with over 200,000 submissions. When I did the basic arithmetic assuming the same infinite processing speed and bandwidth, I found that the pull (ignoring media files) would take around (2,000 id chunks + 200,000 submissions) * 350ms = 70,700,00ms (approximately 19.6 hours).

By introducing concurrency, we can effectively divide the time spent waiting for round trips by the number of concurrent downloads. For example, given 4 independent threads, 19.6 hours turns into 4.9 hours - a significant improvement.

**Comment by yanokwa** [comment]
This is well-explained, @batkinson so big thanks for that! Do you have a sense of how this could be implemented in the current code base?

**Comment by batkinson** [comment]
Hi @yanokwa. Yes, I believe the quickest way would be to change [the submission download loop in ServerFetcher.java](https://github.com/opendatakit/briefcase/blob/master/src/org/opendatakit/briefcase/util/ServerFetcher.java#L314-L334) so that it performs each `downloadSubmission` call on a different thread. You could then further improve on it by making media attachment downloads concurrent in `downloadSubmission` in a similar fashion. First, I would probably just implement it for submissions and measure the speedup using submissions without media files to confirm the approach works and is significantly faster. The biggest thing to be careful of is ensuring that method calls made from `downloadSubmission` are thread-safe.

**Comment by yanokwa** [comment]
Sounds like you've got a good approach! Any chance you've got the cycles to build a first pass and send in a PR?

**Comment by batkinson** [comment]
Sure @yanokwa. I started on a mock-up after writing/running a simulation to test my hypothesis. In case you (or anyone else) is interested, here is what I measured using [the simulation](https://github.com/opendatakit/briefcase/files/1126783/ConcurrentDownload.java.zip). I ran it against example.com from where I'm located (Malabo, Equatorial Guinea).

![performance-graph](https://user-images.githubusercontent.com/478997/27896251-aa8c824a-6210-11e7-877e-678680fdec7f.png)
